Search v1

niccokunzmann opened the pull request:
This branch is based on #34.
It adds the search api compatibility.

niccokunzmann committed:
install the official resource schema package

niccokunzmann committed:
adding resource hook v1

niccokunzmann committed:
Merge branch 'master' into add-resource-hook

niccokunzmann committed:
Merge remote-tracking branch 'origin/master' into add-resource-hook

niccokunzmann committed:
adding service resource-v1

niccokunzmann committed:
add example usage scenarios

niccokunzmann committed:
moved jsonapify hook and installed updtated schema package

niccokunzmann committed:
uploading resource to database

niccokunzmann committed:
add application/vnd.api+json conversion, remove 406 error

niccokunzmann committed:
describe testing

niccokunzmann committed:
split up content type conversion

niccokunzmann committed:
resources can be pushed, stored and received when adding them. step3 works

niccokunzmann committed:
test 5 passed

niccokunzmann committed:
set location header

niccokunzmann committed:
add jsonapi attribute

niccokunzmann committed:
can jsonify error

niccokunzmann committed:
step 18 404 error compatibility

niccokunzmann committed:
add command for elastisearch

niccokunzmann committed:
error code is set, step17

niccokunzmann committed:
/v1/resources/ids

niccokunzmann committed:
test 26 runs - return 204 on no content

niccokunzmann committed:
add content negitiation

niccokunzmann committed:
errors are jsonapi compatible, step29

niccokunzmann committed:
step 30 - not acceptable

niccokunzmann committed:
step 32 passing bad json is a bad requets

niccokunzmann committed:
step 33 convert schema error bad request to unprocessable entity error

niccokunzmann committed:
step 36 running - using originId for queries

niccokunzmann committed:
steps until 36 running. The originId can be used to query a resource

niccokunzmann committed:
step 38: posting wiht same id is 403

niccokunzmann committed:
step 39: invalid ids are detected and return 403

niccokunzmann committed:
step 43 with authentication

niccokunzmann committed:
restricted read access to user, step 44

niccokunzmann committed:
autehntication and access restriction for all hooks

niccokunzmann committed:
test 45 runs

niccokunzmann committed:
resource api tests run

niccokunzmann committed:
first verison of travis build

niccokunzmann committed:
create docker network

niccokunzmann committed:
creating networks

niccokunzmann committed:
logging on failure

niccokunzmann committed:
python install command

niccokunzmann committed:
using pip command

niccokunzmann committed:
using alpne image for faster pull

niccokunzmann committed:
checking for broken service

niccokunzmann committed:
using silent install to shorten the logs

niccokunzmann committed:
Revert "using alpne image for faster pull"

This reverts commit 18feb3ad8a4494f54fe29c8de273153321f08559.

niccokunzmann committed:
another silent install

niccokunzmann committed:
starting twice for connector

niccokunzmann committed:
moved startin steps t install

niccokunzmann committed:
reducing test output if step 1 fails

niccokunzmann committed:
update uuid

https://travis-ci.org/schul-cloud/schulcloud-content/builds/277254875#L2674

niccokunzmann committed:
new package for server tests

niccokunzmann committed:
refactor: put the jsonapi error to the bottom

niccokunzmann committed:
relocate module

https://travis-ci.org/schul-cloud/schulcloud-content/builds/277670184#L1258

niccokunzmann committed:
refactor import to toplevel

niccokunzmann committed:
describe search api usage

niccokunzmann committed:
move errors module

niccokunzmann committed:
use relative path

niccokunzmann committed:
add search endpoint /v1/search

niccokunzmann committed:
test search api

niccokunzmann committed:
moving data test for empty data

niccokunzmann committed:
change order of test installation

niccokunzmann commented:
@kremer-io this can be merged

kremer-io commented:
@niccokunzmann I'll review your PR today

kremer-io commented:
@niccokunzmann This PR also breaking the current implementation of the search service, e.g.:

`GET http://127.0.0.1:4040/search?q="Mathe"`

is returning

```json
{
    "name": "GeneralError",
    "message": "jsonapi-content-type.js: resource can not be converted",
    "code": 500,
    "className": "general-error",
    "data": {},
    "errors": {}
}
```

niccokunzmann committed:
choosing returned content type conversion based on endpoint

niccokunzmann committed:
testing jsonapi result conversion

niccokunzmann commented on scripts/test-content-type-conversion.sh:
```diff
@@ -0,0 +1,71 @@
+assertEndpointIsJsonapi /v1/search
+
+assertEndpointIsNotJsonapi /search
+assertEndpointIsNotJsonapi /resources
```
The script output can be found here: https://travis-ci.org/schul-cloud/schulcloud-content/builds/278114738#L1283

niccokunzmann commented:
An other thing: Can you show me a curl command to post a resource to your endpoint? I will try to find out to test if this works:

- converting resources from the khan crawler to the resource format
- setting userId for khan crawler

niccokunzmann committed:
also testing with mathe and schul as search parameter

niccokunzmann commented:
```
$ curl -X POST http://schulcloud-content-1:content-1@localhost:4040/resources -d '{"originId": "curl123", "providerName": "curl test", "url": "http://localhost:4040/resource", "title":"example test posting", "description": "this is a description", "licenses":["none"], "contentCategory": "atomic", "mimeType": "applcation/json" }'
{
  "name": "BadRequest",
  "message": "Data does not match schema",
  "code": 400,
  "className": "bad-request",
  "data": {},
  "errors": [
    "should have required property 'title'",
    "should have required property 'description'",
    "should have required property 'url'",
    "should have required property 'contentCategory'",
    "should have required property 'mimeType'",
    "should have required property 'licenses'"
  ]
}
```

niccokunzmann commented:
Ah, I get it it is in the jsonapify hook. Hab ich übersehen.

niccokunzmann commented:
habs:

```
curl -X POST -H "Content-Type: application/vnd.api+json" http://schulcloud-content-1:content-1@localhost:4040/resources --data '{"data":{"attributes":{"originId": "curl123", "providerName": "curl test", "url": "http://localhost:4040/resource", "title":"example test posting", "description": "this is a description", "licenses":["none"], "contentCategory": "atomic", "mimeType": "applcation/json" }}}'
```

niccokunzmann commented:
Kannst du mir ein valides curl-commando geben, das deine Suche anfragt?

niccokunzmann committed:
implemented crossposting resources from different apis

kremer-io commented:
z.B. `curl http://127.0.0.1:4040/search?q=Mathe`

niccokunzmann commented:
Was ist dein output?

niccokunzmann commented:
gerade bekomme ich abgehacktes JSON: https://travis-ci.org/schul-cloud/schulcloud-content/builds/278150518#L1436

niccokunzmann committed:
change order to also test other apis search result ceonversion

niccokunzmann committed:
using spaces to return full data

niccokunzmann committed:
output of failed response

niccokunzmann commented:
Fehler: https://travis-ci.org/schul-cloud/schulcloud-content#L1352

niccokunzmann commented:
https://travis-ci.org/schul-cloud/schulcloud-content/builds/278159183#L1352

niccokunzmann committed:
try to remove general error

    Model,
https://travis-ci.org/schul-cloud/schulcloud-content/builds/278159183#L1352

niccokunzmann committed:
Revert "try to remove general error"

This reverts commit 869773caf9f66f18a3144062d25003d2b415eee0.

niccokunzmann committed:
set Model of class Service for jsonapi

niccokunzmann committed:
remove other tests to see error

niccokunzmann committed:
document hw to rebuild

niccokunzmann committed:
document how to get the output

niccokunzmann committed:
restore old service class

niccokunzmann committed:
change order of app registration

niccokunzmann commented:
Mein Problem ist gerade, dass ich /search nicht angefasst habe, es aber trotzdem abstürzt: 

https://github.com/schul-cloud/schulcloud-content/pull/36/files#diff-19b0242cfe04c92831da1b7542ddbb0b

niccokunzmann commented:
```
{

  "name": "GeneralError",

  "message": "Cannot read property 'modelName' of undefined",

  "code": 500,

  "className": "general-error",

  "data": {},

  "errors": {}

}
```

und 

```
schulcloud-content_1                    |0m info: error: search - Method: find: Cannot read property 'modelName' of undefined

schulcloud-content_1                    | error:  TypeError: Cannot read property 'modelName' of undefined

schulcloud-content_1                    |     at module.exports (/app/src/hooks/jsonapify/find.js:14:37)

schulcloud-content_1                    |     at Object.<anonymous> (/app/src/hooks/jsonapify/index.js:68:9)

schulcloud-content_1                    |     at process._tickCallback (internal/process/next_tick.js:109:7)
```

niccokunzmann committed:
remove jsonapify hook from /search

niccokunzmann committed:
add other tests again

niccokunzmann commented:
Closing this to create a pull-requests into the development branch.
